File: src/io/innofang/singleton/example/java/ThreadSafeDoubleCheckSingleton.java
package io.innofang.singleton.example.java;

/**
 * Created by Inno Fang on 2017/8/12.
 */
public class ThreadSafeDoubleCheckSingleton {

    private static volatile ThreadSafeDoubleCheckSingleton instance;

    private ThreadSafeDoubleCheckSingleton() {}

    public static ThreadSafeDoubleCheckSingleton getInstance() {
        if (null == instance) {
            synchronized (ThreadSafeDoubleCheckSingleton.class) {
                if (null == instance) {
                    instance = new ThreadSafeDoubleCheckSingleton();
                }
            }
        }
        return instance;
    }

}

File: src/io/innofang/singleton/example/kotlin/ThreadSafeDoubleCheckSingleton.kt
package io.innofang.singleton.example.kotlin

/**
 * Created by Inno Fang on 2017/8/12.
 */
class ThreadSafeDoubleCheckSingleton {
    companion object {
        // implement 1
        val instance1 by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            ThreadSafeDoubleCheckSingleton()
        }

        // implement 2
        @Volatile private var instance2: ThreadSafeDoubleCheckSingleton? = null

        fun get(): ThreadSafeDoubleCheckSingleton {
            if (null == instance2) {
                synchronized(this) {
                    if (null == instance2) {
                        instance2 = ThreadSafeDoubleCheckSingleton()
                    }
                }
            }
            return instance2!!
        }
    }
}